File: backend/app/schemas/mechanism.py
from pydantic import BaseModel
from datetime import datetime
from typing import List, Optional

from backend.app.schemas.user import UserResponse

# カテゴリー情報表示用スキーマ
class CategoryResponse(BaseModel):
    id: int
    name: str

    class Config:
        from_attributes = True

# メカニズム作成用スキーマ
class MechanismCreate(BaseModel):
    title: str
    description: str
    reliability: int
    categories: List[str]

# メカニズム編集用スキーマ
class MechanismUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    reliability: Optional[int] = None
    categories: Optional[List[str]] = None

# メカニズム情報表示用スキーマ（一覧用）
class MechanismListResponse(BaseModel):
    id: int
    title: str
    description: str
    reliability: int
    thumbnail_path: Optional[str] = None
    user: UserResponse
    categories: List[str]
    likes_count: int
    views_count: Optional[int] = 0
    created_at: datetime

    class Config:
        from_attributes = True

# メカニズム情報表示用スキーマ（詳細用）
class MechanismDetailResponse(BaseModel):
    id: int
    title: str
    description: str
    reliability: int
    file_path: str
    thumbnail_path: Optional[str] = None
    user: UserResponse
    categories: List[str]
    likes_count: int
    views_count: Optional[int] = 0
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

# ページネーション用スキーマ
class PaginatedMechanismResponse(BaseModel):
    items: List[MechanismListResponse]
    total: int
    page: int
    limit: int
    pages: int

File: frontend/src/types/mechanism.ts
// メカニズム関連の型定義

import { User } from './auth';

// メカニズム基本情報の型
export interface Mechanism {
  id: number;
  title: string;
  description: string;
  reliability: number;
  thumbnail_path: string | null;
  user: User;
  categories: string[];
  likes_count: number;
  views_count?: number;  // 閲覧回数（オプション）
  created_at: string;
}

// メカニズム詳細情報の型（基本情報を拡張）
export interface MechanismDetail extends Mechanism {
  file_path: string;
  updated_at: string;
}

// メカニズム一覧レスポンスの型
export interface PaginatedMechanismResponse {
  items: Mechanism[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

// メカニズム作成フォームの型
export interface MechanismFormData {
  title: string;
  description: string;
  reliability: number;
  categories: string;
  file: File | null;
  thumbnail: File | null;
}

// メカニズム閲覧回数の型
export interface MechanismViewCount {
  mechanism_id: number;
  total_views: number;
  user_views?: number;
}

// 複数メカニズムの閲覧回数レスポンスの型
export interface MechanismViewsResponse {
  items: MechanismViewCount[];
}
